File: src/main/kotlin/com/github/dungphan/unityindex/tools/unity/UnityGraphSnapshotTool.kt
package com.github.dungphan.unityindex.tools.unity

import com.github.dungphan.unityindex.constants.ToolNames
import com.github.dungphan.unityindex.server.models.ToolCallResult
import com.github.dungphan.unityindex.tools.AbstractMcpTool
import com.github.dungphan.unityindex.tools.models.GraphSnapshotRequest
import com.github.dungphan.unityindex.tools.schema.SchemaBuilder
import com.github.dungphan.unityindex.util.UnityAssetGraphBuilder
import com.intellij.openapi.project.Project
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Day 2 MCP surface for the Unity asset graph. Wire format documented in
 * `docs/graph-mcp-tools.md` §3.1 (request/response) and `docs/graph-schema.md`
 * (node + edge taxonomy).
 *
 * Asset-domain only. Code edges (csharp nodes) arrive in Day 8.
 */
class UnityGraphSnapshotTool : AbstractMcpTool() {

    override val requiresPsiSync: Boolean = false

    override val name: String = ToolNames.UNITY_GRAPH_SNAPSHOT

    override val description: String = """
        Return the full Unity asset graph as a GraphSnapshot — every script, prefab, scene, ScriptableObject, and asset under the project, plus the edges between them (script_used_by_prefab/scene, scene_contains_prefab, prefab_variant_of, serialized_binding, script_declares_class).

        Sub-file kinds (`component_instance`, `component_field`) are never returned as top-level nodes; their counts go into `stats.skipped_component_*` and the underlying IDs ride along as edge metadata. Use `unity_graph_expand` (Phase 1, ships Day 6 or 7) to materialize them for a single container.

        `script_declares_class` edges point to `unity://csharp/T:<ClassName>` IDs that Day 8 (`unity_graph_code_edges`) will materialize. A single `dangling_csharp_targets` warning is emitted when any such edges are present.

        Parameters:
        - include_kinds (optional): NodeKind[] — keep only nodes of these kinds.
        - exclude_kinds (optional): NodeKind[] — applied after include_kinds.
        - path_globs (optional): string[] — include-only path filter (** any depth, * single segment, ? single char).
        - include_orphans (optional): bool — default true; when false, drop degree-0 nodes.
        - pagination (optional): { page_size?: int, cursor?: string } — opaque cursor.
        - project_path (optional): only needed when multiple projects are open.
    """.trimIndent()

    override val inputSchema: JsonObject = SchemaBuilder.tool()
        .projectPath()
        .property("include_kinds", buildJsonObject {
            put("type", JsonPrimitive("array"))
            put("description", JsonPrimitive("Restrict the snapshot to these NodeKinds (e.g. script, prefab, scene, so, asset)."))
            put("items", buildJsonObject {
                put("type", JsonPrimitive("string"))
            })
        })
        .property("exclude_kinds", buildJsonObject {
            put("type", JsonPrimitive("array"))
            put("description", JsonPrimitive("Drop nodes of these NodeKinds. Applied after include_kinds."))
            put("items", buildJsonObject { put("type", JsonPrimitive("string")) })
        })
        .property("path_globs", buildJsonObject {
            put("type", JsonPrimitive("array"))
            put("description", JsonPrimitive("Project-relative globs that nodes must match; edges crossing the boundary are dropped."))
            put("items", buildJsonObject { put("type", JsonPrimitive("string")) })
        })
        .booleanProperty("include_orphans", "Default true. When false, nodes with degree 0 are dropped.")
        .property("pagination", buildJsonObject {
            put("type", JsonPrimitive("object"))
            put("description", JsonPrimitive("Opaque pagination cursor. Slice nodes; edges crossing the page boundary are dropped."))
            put("properties", buildJsonObject {
                put("page_size", buildJsonObject {
                    put("type", JsonPrimitive("integer"))
                    put("description", JsonPrimitive("Default 5000, max 20000."))
                })
                put("cursor", buildJsonObject {
                    put("type", JsonPrimitive("string"))
                    put("description", JsonPrimitive("Opaque cursor from the previous response."))
                })
            })
        })
        .stringProperty("request_id", "Optional; echoed back on the response for client correlation.")
        .build()

    override suspend fun doExecute(project: Project, arguments: JsonObject): ToolCallResult {
        val request = try {
            json.decodeFromJsonElement(GraphSnapshotRequest.serializer(), arguments)
        } catch (e: Exception) {
            return createErrorResult("Invalid arguments: ${e.message}")
        }

        return try {
            // UnityAssetGraphBuilder touches VFS + parses raw YAML — no PSI.
            // Wrapping in a platform read action held the read lock through a
            // multi-minute walk on large projects, starving write-intent actions
            // and freezing the EDT. Run on IO instead.
            val response = withContext(Dispatchers.IO) {
                UnityAssetGraphBuilder.build(project, request)
            }
            createJsonResult(response)
        } catch (e: IllegalStateException) {
            createErrorResult(e.message ?: "Failed to build asset graph")
        } catch (e: Exception) {
            createErrorResult("Failed to build asset graph: ${e.message}")
        }
    }

}

File: vscode-extension/src/tools/unity/unityGraphSnapshotTool.ts
import type { SnapshotRequest, SnapshotResponse } from "@unity-index/graph-core";
import { materializeClassAnchors } from "@unity-index/graph-core";
import { AbstractMcpTool, ToolContext } from "../abstractTool";
import { TOOL_NAMES } from "../../constants";
import { ToolCallResult } from "../../models/jsonRpc";
import { ProjectContext } from "../../server/projectResolver";
import { Args } from "../../utils/args";
import { SchemaBuilder } from "../../utils/schema";

/**
 * Day 2 MCP surface for the Unity asset graph. Wire format documented in
 * docs/graph-mcp-tools.md §3.1 and docs/graph-schema.md.
 *
 * Asset-domain only. Code edges (csharp nodes) arrive in Day 8.
 */
export class UnityGraphSnapshotTool extends AbstractMcpTool {
  protected readonly requiresLsp = false;
  readonly isHeavyScan = true;

  readonly name = TOOL_NAMES.UNITY_GRAPH_SNAPSHOT;
  readonly description =
    "Return the full Unity asset graph as a GraphSnapshot — every script, prefab, scene, ScriptableObject, and asset under the project, plus the edges between them (script_used_by_prefab/scene, scene_contains_prefab, prefab_variant_of, serialized_binding, script_declares_class). " +
    "Sub-file kinds (component_instance, component_field) are never returned as top-level nodes; their counts go into stats.skipped_component_* and the underlying IDs ride along as edge metadata. Use unity_graph_expand (Phase 1, ships Day 6 or 7) to materialize them for a single container. " +
    "script_declares_class edges point to unity://csharp/T:<ClassName> IDs that Day 8 (unity_graph_code_edges) will materialize. A single dangling_csharp_targets warning is emitted when any such edges are present (set include_class_anchors=true to suppress).";
  readonly inputSchema = SchemaBuilder.tool()
    .projectPath()
    .property("include_kinds", {
      type: "array",
      description:
        "Restrict the snapshot to these NodeKinds (e.g. script, prefab, scene, so, asset).",
      items: { type: "string" },
    })
    .property("exclude_kinds", {
      type: "array",
      description: "Drop nodes of these NodeKinds. Applied after include_kinds.",
      items: { type: "string" },
    })
    .property("path_globs", {
      type: "array",
      description:
        "Project-relative globs that nodes must match; edges crossing the boundary are dropped.",
      items: { type: "string" },
    })
    .booleanProperty(
      "include_orphans",
      "Default true. When false, nodes with degree 0 are dropped.",
    )
    .booleanProperty(
      "include_class_anchors",
      "Default false. When true, materialize one `class` node per `script_declares_class` edge target (anchor=true) so the UI has stable IDs to attach Day 8 code edges to. Suppresses the dangling_csharp_targets warning.",
    )
    .property("pagination", {
      type: "object",
      description:
        "Opaque pagination cursor. Slices nodes; edges crossing the page boundary are dropped.",
      properties: {
        page_size: {
          type: "integer",
          description: "Default 5000, max 20000.",
        },
        cursor: {
          type: "string",
          description: "Opaque cursor from the previous response.",
        },
      },
    })
    .stringProperty(
      "request_id",
      "Optional; echoed back on the response for client correlation.",
    )
    .build();

  /** Day 8.4 — opt-in projection applied AFTER the cache so the cached
   *  unfiltered snapshot stays anchor-free and we don't bloat cold reads
   *  that don't ask for them. Pure function: `response` is not mutated. */
  private applyClassAnchors(
    response: SnapshotResponse,
    request: SnapshotRequest,
  ): SnapshotResponse {
    if (!request.include_class_anchors) return response;
    const result = materializeClassAnchors(response.snapshot, {
      warnings: response.warnings,
    });
    if (result.anchorsAdded === 0) return response;
    return {
      ...response,
      snapshot: result.snapshot,
      warnings: result.warnings,
    };
  }

  protected async doExecute(
    project: ProjectContext,
    args: Args,
    ctx: ToolContext,
  ): Promise<ToolCallResult> {
    const request = args as unknown as SnapshotRequest;
    try {
      const index = await ctx.assetIndex.get(project);
      const response = await ctx.graphCache.getSnapshot(
        project,
        index,
        request,
        ctx.signal,
      );
      return this.json(this.applyClassAnchors(response, request));
    } catch (e) {
      const message = e instanceof Error ? e.message : String(e);
      return this.error(`Failed to build asset graph: ${message}`);
    }
  }
}
